Training24/intlist.c

```diff
@@ -0,0 +1,179 @@
+// ------------------------------------------------------------------------------------------------
+// Training ~ A training program for new joiners at Metamation, Batch - July 2024.
+// Copyright (c) Metamation India.
+// ------------------------------------------------------------------
+// intlist.c
+// Program on Linked list functions.
+// ------------------------------------------------------------------------------------------------
+#include <malloc.h>
+#include "intlist.h"
+
+int mlastError = SUCCESS;  // Global variable to store the last error code
+
+IntList* Create () {
+   IntList* newList = (IntList*)malloc (sizeof (IntList));
+   if (newList == NULL) {
+      mlastError = MEMORY_ALLOCATION_FAILED;
+      return NULL;
+   }
+   newList->head = NULL; ; // Intialilze head pointer as null
+   mlastError = SUCCESS;
+   return newList;
+}
+
+void Delete (IntList* list) {
+   if (list == NULL) {
+      mlastError = NULL_POINTER;
+      return;
+   }
+   Node* current = list->head;
+   Node* next;
+   while (current != NULL) {  // Traverse the list and free each node
+      next = current->next;
+      free (current);
+      current = next;
+   }
+   free (list);
+   mlastError = SUCCESS;
+}
+
+void Add (IntList* list, int data) {
+   if (list == NULL) {
+      mlastError = NULL_POINTER;
+      return;
+   }
+   Node* newNode = (Node*)malloc (sizeof (Node));
+   if (newNode == NULL) {
+      mlastError = MEMORY_ALLOCATION_FAILED;
+      return;
+   }
+   newNode->data = data;
+   newNode->next = NULL;
+   if (list->head == NULL) list->head = newNode;    //If the list is empty, set the new node as the head
+   else {
+      Node* current = list->head;
+      while (current->next != NULL) {  // Traverse to the end of the list
+         current = current->next;
+      }
+      current->next = newNode;
+   }
+   mlastError = SUCCESS;
+}
+
+void Insert (IntList* list, int index, int data) {
+   if (list == NULL) {
+      mlastError = NULL_POINTER;
+      return;
+   }
+   if (index < 0 || index > Count (list)) {
+      mlastError = INVALID_INDEX;
+      return;
+   }
+   Node* newNode = (Node*)malloc (sizeof (Node));
+   if (newNode == NULL) {
+      mlastError = MEMORY_ALLOCATION_FAILED;
+      return;
+   }
+   newNode->data = data;
+   if (index == 0) { // If inserting at the head
+      newNode->next = list->head;
+      list->head = newNode;
+   }
+   else {
+      Node* current = list->head;
+      for (int i = 0; i < index - 1; i++) current = current->next;   // Traverse to the node before the insertion point

```

Comment:
Here the loop has to iterate till it reaches the node before the specified index. But the condition stops it before reaching the node. Please check it.